### PR TITLE
Add ExplodedNode.ToString() for debugging SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #5306

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs [hunk 1]
```diff
@@ -62,5 +62,10 @@ private ExplodedNode(BasicBlock block, IOperationWrapperSonar[] operations, int
             other is not null
             && other.programPointHash == programPointHash
             && other.State.Equals(State);
+
+        public override string ToString() =>
```
Comment (Corniel) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs:66]: For debugging, newlines are not particular useful. Consider to (also) have a [DebuggerDisplay] attribute that has a single line representation.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs:66]: Hi @Corniel, thanks for the suggestion. We've considered the `[DebuggerDisplay]` option and went for the `ToString`. While the `ExplodedNode` itself doesn't really need the new line, the `ProgramState` does.

The intended usage is to add `Console.WriteLine(current.ToString())` locally here, as the flows can be often tricky to follow and understand:
https://github.com/SonarSource/sonar-dotnet/blob/3f1f4cdbebba4d87acd571579389bb7d8324a9ed/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs#L50-L62

Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs:66]: You could have a debugger display to, that does `ToString().Replace(Environement.NewLine, "; "). For Console.WriteLine, new lines indeed have added value.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs:66]: We'll see if that's needed later. For now, I'm a bit afraid that the state will be too long to read in a single line anyway based on the experience with the old engine.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ExplodedNode.cs:66]: But I like the `Replace` idea if we'll need taht

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs [hunk 2]
```diff
@@ -66,5 +68,34 @@ private ProgramState()
...
+                if (OperationValue.Any())
+                {
+                    sb.AppendLine("Operations:");
+                    foreach (var kvp in OperationValue.Select(x => new KeyValuePair<string, string>(x.Key.Serialize(), x.Value?.ToString() ?? "<null>")).OrderBy(x => x.Key))
+                    {
+                        sb.Append(kvp.Key).Append(": ").AppendLine(kvp.Value);
+                    }
+                }
+                return sb.ToString();
+            }
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:98]: I think we should extract 2 methods: GetSymbols, GetOperations and then transform this full method into an expression.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExplodedNodeTest.cs [hunk 3]
```diff
@@ -131,5 +131,23 @@ public void GetHashCode_ReturnsSameForEquivalent()
             basic.GetHashCode().Should().NotBe(differentLocation.GetHashCode());
             basic.GetHashCode().Should().NotBe(differentState.GetHashCode());
         }
+
+        [TestMethod]
+        public void ToString_SerializeOperationAndState()
+        {
+            var cfg = TestHelper.CompileCfgBodyCS("var a = true;");
+            var state = ProgramState.Empty.SetSymbolValue(cfg.Blocks[1].Operations[0].Children.First().TrackedSymbol(), new(new()));
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExplodedNodeTest.cs:139]: `new(new())` -> hard to figure out what's happening here without looking at the method signature.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ProgramStateTest.cs [hunk 4]
```diff
@@ -315,6 +316,88 @@ public void GetHashCode_ReturnsSameForEquivalent()
...
+a: SV_1
+Sample.Main(): SV_2: Second
+");
+        }
+
+        [TestMethod]
+        public void ToString_WithOperations()
+        {
+            var assignment = TestHelper.CompileCfgBodyCS("var a = true;").Blocks[1].Operations[0];
+            var counter = new SymbolicValueCounter();
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ProgramStateTest.cs:355]: We use a counter in all the tests. We could make it a field and use a TestInitialize method to construct it before each test.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExplodedNodeTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ExplodedNodeTest.cs:139]: That's a kind of build-in for target typed new feature. As we're inside UT, it is just a boring scaffolding.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ProgramStateTest.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ProgramStateTest.cs:355]: I'm considering refactoring these UTs to use `TestInitialize` to generate some `context` field that would provide `Operation1`, `Operation2`, `Operation3`, `Value1`, `Value2`, `Value3`, `Symbol1`, `Symbol2` and `Symbol3` as those repeat in most of the cases. The same type could be then used in `ProgramStateTest`. In another PR

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs [hunk 5]
```diff
@@ -66,5 +68,32 @@ private ProgramState()
             other is not null
             && other.OperationValue.DictionaryEquals(OperationValue)
             && other.SymbolValue.DictionaryEquals(SymbolValue);
+
+        public override string ToString() =>
+            Equals(Empty) ? "Empty" : SerializeSymbols() + SerializeOperations();
+
+        private string SerializeSymbols() =>
+            Serialize(SymbolValue, "Symbols", x => x.ToString(), x => x?.ToString() ?? "<null>");
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs:76]: This last part `x => x?.ToString() ?? "<null>"`, can be included in the method.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! Just a small comment to address.

**Pull request merged**